### PR TITLE
Globus integration

### LINKED ISSUE #1188: Upload via Globus button
The title for this ticket is deceptively simple. It involves frontend JavaScript work, backend database work, and API interactions.

The page to create a new Project (the home page) should include a button "Request Globus link".  When clicked a new upload directory in the Globus Endpoint should be created for the user, and the user's browser should open a new tab pointed to the Globus File Manager for that directory opened as a new tab in their browser, and the *Staging Location* in the form should be filled in with the URL for the new Globus location.

Screenshot 2023-09-07 at 7 07 36 AM

The logic for creating the upload directory is similar to [what is in H2](https://github.com/sul-dlss/happy-heron/blob/main/app/jobs/globus_setup_job.rb#L35-L46) but we are going to handle the request synchronously (not part of a Sidekiq job), and we are going to use a timestamp instead of a Work ID for the directory.

When a user clicks *Request Globus link*:

1. An HTTP POST to the server will be sent indicating the user wants a Globus directory.
2. The receiving Rails controller will create a new `GlobusDestination` (a new AR model) that records the `user_id`, `created_at` and `deleted_at` (which is null).
3. The controller will create the Globus upload directory for the user using `GlobusClient.mkdir` with a `path` of `/{sunet}/{GlobusDestination.created_at}/}`. The `GlobusClient` should be configured using Settings similar to [what is in H2](https://github.com/sul-dlss/happy-heron/blob/main/config/settings.yml#L10-L14).
4. The Rails controller will tell the browser to update the *Staging Location* input box with the URL for the new Globus endpoint, e.g. https://app.globus.org/file-manager?&destination_id=f3e29605-7ba5-45f5-8900-f721663cf1af&destination_path=/edsu/20230907044801/, and will also open a new tab pointed at the Globus File Manager for the new folder.  It may be useful to add a `GlobusDestination.url`, `GlobusDestination.destination_path` and `GlobusDestination.staging_location` helper methods to isolate the logic for these.

When the user finally submits the form to create the new Project aka BatchContext, it is linked (one-to-one) with the appropriate `GlobusDestination`. The controller will need to be adjusted to be able to: 

1. Allow for the form field to be a Globus URL
2. Use the `destination_path` in the Globus URL and the current user to look up a `GlobusDestination` object.
3. Set the `BatchContext.staging_location` to the full path to the staging location on the server (using a `GlobusDestination.staging_location` helper method).
4. Persist the new `BatchContext` and updated `GlobusDestination`
5. Continue JobRun as normal

To prevent unintentional creation of too many access points (Globus imposes a max of 100) if the user has 99 active GlobusDestination records associated with their account (`deleted_at` = nil) the button will be disabled and an error message will be displayed: 

Screenshot 2023-09-05 at 12 08 41 PM

The message should read "You have too many active please email SDR Manager for help" which links to an email alias for the SDR manager.

The deletion of access rules will be handled in #1189. The use of access rules to manage access will allow us to be able to mark a directory as processed, which can aid in deleting data from the preassembly mounts in the future.

### PR DESCRIPTION
**Pull request by edsu**
# Why was this change made? 🤔

Introduces importmap-rails and stimulus-rails to add Request Globus Link functionality to the Project create page. Once clicked this will create a Globus directory with access rule set up for the current Preassembly user. The URL for the Globus Viewer is added to the form as the Staging Location. When submitted the GlobusDestination is linked up with the BatchContext.

Fixes #1188

# How was this change tested? 🤨

- [X] CI
- [x] Integration tests

# Does your change introduce accessibility violations? 🩺

⚡ ⚠ Please ensure this change does not introduce accessibility violations (at the WCAG A or AA conformance levels); if it does, include a rationale. See the [Infrastructure accessibility guide](https://github.com/sul-dlss/DeveloperPlaybook/blob/main/best-practices/infra-accessibility.md) for more detail. ⚡

TBD

### CODE REVIEW

File: app/controllers/batch_contexts_controller.rb
Comment (peetucket) [inline app/controllers/batch_contexts_controller.rb:27]: There is some validation around `staging_location`: https://github.com/sul-dlss/pre-assembly/blob/main/app/models/batch_context.rb#L155-L176

Does it work as expected with a globus URL in that field?

Comment (edsu) [reply app/controllers/batch_contexts_controller.rb:27]: Yes, but only because that while submitted with the staging_location form field, it is swapped out for the actual location prior to the object getting created. See https://github.com/sul-dlss/pre-assembly/blob/t1188-form/app/controllers/batch_contexts_controller.rb#L21-L26

File: app/controllers/globus_controller.rb
Comment (edsu) [inline app/controllers/globus_controller.rb:34]: This is the new controller that the Stimulus JavaScript component POSTs to when they click to "Request Globus Link"

File: app/javascript/controllers/globus_controller.js
Comment (edsu) [inline app/javascript/controllers/globus_controller.js:42]: This is the new Stimulus Controller that contacts the server when the user clicks the "Request Globus Link" button. Once clicked the button immediately deactivates (to prevent repetitive clicking) and then turns into a hyperlink once the Globus location has been created. The Globus File viewer is opened in a new tab, and the Staging Location is populated with the Globus URL.

File: app/models/batch_context.rb
Comment (edsu) [inline app/models/batch_context.rb:18]: I think it makes more sense for this to be singular?

File: app/views/batch_contexts/_new_bc_form.erb
Comment (edsu) [inline app/views/batch_contexts/_new_bc_form.erb:11]: The new Request Globus Link button. Note I needed to enclose the whole form in a div in order to set the StimulusController on it, which allows it to update the Staging Location after the Globus Link was created.

File: app/views/layouts/application.html.erb
Comment (edsu) [inline app/views/layouts/application.html.erb:13]: This should be handled by the importmap now.

File: db/migrate/20230928211022_add_directory_to_globus_destination.rb
Comment (edsu) [inline db/migrate/20230928211022_add_directory_to_globus_destination.rb:17]: Because serializing and parsing dates with micro-seconds was a bit tedious I ended up adding the Directory field to the GlobusDeposit model. This makes it easier to set and get it without needing to infer it from the created_at timestamp. It also means we can easily change the default directory name in the future if we want.

File: db/schema.rb
Comment (edsu) [inline db/schema.rb:13]: I'm not entirely sure how this happened. Hopefully it won't cause a problem for others. I did deploy to QA without a problem.

**Comment by edsu** [comment]
Testing by @andrewjbtw revealed that Globus is sending emails when the Access Rule is created for the user (when they click on Create Globus Link). However that email has a link in it that flips the origin & destination such that the `destination_path` parameter is not present (see also https://github.com/sul-dlss/happy-heron/issues/3374). For example:

https://app.globus.org/file-manager?&origin_id=75347226-3844-4d81-ba64-bba202f205c1&origin_path=/edsu/2023-10-02-10-41-39-534/&add_identity=d28a330f-9d3c-4832-950c-492fb7771a9e

We could simply pull use the `origin_path` when `destination_path` is not, but when a user navigates to a directory that they want to transfer from that will then populate the `destination_path` and `destination_id` parameters.  For example:

https://app.globus.org/file-manager?add_identity=d28a330f-9d3c-4832-950c-492fb7771a9e&destination_id=96bd6268-e43f-11ed-ba2f-09d6a6f08166&destination_path=%2F~%2F&origin_id=75347226-3844-4d81-ba64-bba202f205c1&origin_path=%2Fedsu%2F2023-10-02-10-41-39-534%2F

So to reliably detect the GlobusEndpoint to look up we should:

1. See if `destination_id` matches the Globus Destination that the app is configured for `Settings.globus.endpoint_id` in which case use the `destination_path`.
2. See if `origin_id` matches `Settings.globus.endpoint_id` in which case use `origin_path` instead.
3. If neither 1 or 2 then allow an error about it not being a recognizable Staging Location.

File: app/models/batch_context.rb
Comment (lwrubel) [reply app/models/batch_context.rb:18]: Yeah, that was my bad, thanks for fixing. I think Rails still did the right thing(?), but should be singular.

**Comment by edsu** [comment]
I moved improving the detection of Globus URLs to a separate ticket: #1333

File: app/javascript/controllers/globus_controller.js
Comment (ndushay) [reply app/javascript/controllers/globus_controller.js:42]: could this be a comment in the code?

File: app/models/globus_destination.rb
Comment (ndushay) [reply app/models/globus_destination.rb:40]: this could also be a comment in the code.

File: app/javascript/controllers/globus_controller.js
Comment (peetucket) [inline app/javascript/controllers/globus_controller.js:27]: small typo `witht` -> `with`

File: db/schema.rb
Comment (ndushay) [reply db/schema.rb:13]: when you have a db/migrate, it creates a new db schema with a new datestamp.  Kinda like Gemfile.lock or .rubocop_todo.yml

File: app/javascript/controllers/globus_controller.js
Comment (peetucket) [inline app/javascript/controllers/globus_controller.js:19]: does throwing the error show anything to the user ?

File: spec/models/globus_destination_spec.rb
Comment (ndushay) [inline spec/models/globus_destination_spec.rb:7]: ooo - I like this better - easier to grok, IMO.

**Review by ndushay — APPROVED** [review]
LGTM;  a couple quibbles.

Also:  
- is it true andrew signed off?  no evidence in the PR comments
- maybe squash the commits?

File: db/schema.rb
Comment (peetucket) [reply db/schema.rb:13]: I don't think this has an impact on anything so no need to worry about it ... the scema.rb is for documentation and to re-create the database without migrations if needed.

**Review by peetucket — APPROVED** [review]

File: app/models/globus_destination.rb
Comment (edsu) [reply app/models/globus_destination.rb:40]: It is a comment already?

File: app/javascript/controllers/globus_controller.js
Comment (edsu) [reply app/javascript/controllers/globus_controller.js:19]: It shows that it is an invalid staging location.

Comment (edsu) [reply app/javascript/controllers/globus_controller.js:19]: Oh actually it doesn't -- sorry I was initially confused and thinking this was in the Rails controller. I can have it throw an alert, which won't be the prettiest but at least will get noticed.

Comment (edsu) [reply app/javascript/controllers/globus_controller.js:19]: I misread your comment initially and thought this was about the Rails controller, not the Stimulus one! If it's ok, I'd like to address this in a follow on ticket #1340  to move this work along, since it is holding up other work.

**Comment by edsu** [comment]
Ugh, I forgot to squash commits before merging -- my apologies.

**Pull request merged**
